Replace span relationship with a potentially remote parent context

krnowak opened the pull request:
This PR removes the non-compliant ChildOf and FollowsFrom interfaces
and the Relation type, which were inherited from OpenTracing via the
initial prototype. The logic I propose is here, which says to use the
remote span context if its TraceID differs from the current Span.

The opentracing bridge will likely need to see more work, but I'd like to punt it to the separate PR later.

This PR was a part of #381.

MrAlias commented on api/trace/context.go:
```diff
@@ -0,0 +1,24 @@
+
+// RemoteContext gets the current core.SpanContext from a Context.
+func RemoteContext(ctx context.Context) core.SpanContext {
+	if sc, ok := ctx.Value(remoteContextKey).(core.SpanContext); ok {
```
My assertion this will cause a panic is incorrect. Forget I made this suggestion.

rghetia commented on api/trace/api.go:
```diff
@@ -189,6 +174,26 @@ func (sk SpanKind) String() string {
+//     a. If a current span is set with the same trace_id, use the span's
+//        SpanContext as parent.
+//     b. If the current span belongs to a different trace_id, use the remote
+//        context as parent.
```
Is condition 1.b tested? 
I looked at the tests but didn't find one. Maybe it is there and I just missed it.

jmacd replied:
I had also proposed that the current trace_id be used to decide whether to take the current span or the remote context, and in that discussion was convinced (by @mwear) that we should do the simpler thing. Always use the span as parent if there is one, then fall back to the remote. It means that, in some cases, we'll have to clear the current span in order to use the remote context, which I think is acceptable.  See here: https://github.com/open-telemetry/opentelemetry-specification/pull/423

krnowak commented:
The longer I look at this PR, the more convinced I become that `WithParent` is not a way to go. I feel uneasy when I see that the tracer may need to somehow deal with two contexts (one passed as a first parameter to the `Start` function, and one potentially passed with the `WithParent` option). As it is in the PR currently, when `WithParent` is used, we completely ignore the context passed explicitly to the `Start` function. This may be confusing if the Otel user had code like:

```go
func doStuff(ctx context.Context, …) {
	someStuff := getSomeStuff()
	ctx = context.WithValue(ctx, myStuffKey, someStuff)
	ctx, span := getTracer().Start(ctx, WithParent(WithRemoteParent(context.Background(), conjureSomeSpanContext())))
	…
	// much, much later, somewhere deeper in the call stack
	// panic, ctx is a background one with some span context
	someStuff := ctx.Value(myStuffKey).(stuffType)
	…
}
```

So far, we avoided that pitfall in examples and tests, because we usually passed the same context as a first parameter to `Start` and to the `WithParent` option, or we passed a throw-away context to `Start` and a useful context to `WithParent`, like:

```go
ctx, span := tracer.Start(ctx, WithParent(WithRemoteParent(ctx, someSC)))
ctx2, span2 := tracer.Start(context.Background(), WithParent(WithRemoteParent(ctx, someSC)))
```

This is just lame, IMO.

I'd propose the following:

- drop WithParent option
- add functions for clearing current span and remote span context from context (just in case someone really wants a new root)

I'll try to implement that and also see how opentracing bridge can cope with it.

rghetia commented:
The clearing of remote span context can be implicit with option WithNewRoot. 
```go
	ctx, span := getTracer().Start(ctx, WithNewRoot())
```
This option would create a new span and add link to remote context and existing span context.
If this option is not present then
use SpanContext as parent.
use RemoteSpanContext as parent if SpanContext is not present
use new root if both SpanContext and RemoteSpanContext are not present.

jmacd commented:
Although I wrote the `WithParent()` option in the earlier, I think I agree with @krnowak that we shouldn't keep it. Independent of the `WithNewRoot()` proposal here, the user can just pass one context, either the extracted context to use the remote parent, or some other context to use the current span.

krnowak commented:
Updated. No more `WithParent`, got the `WithNewRoot` option. Fixed opentracing bridge, but otel spans created with the opentracing bridge are always going to have a remote parent. Tough luck.

rghetia commented on internal/trace/parent/parent.go:
```diff
@@ -0,0 +1,41 @@
+	return append(links, trace.Link{
+		SpanContext: sc,
+		Attributes: []core.KeyValue{
+			key.String("ignored-on-demand", kind),
```
I am not sure about `ignored-on-demand`.  kind (key/value) should be defined in sdk-spec.
How about 'link-type'?

krnowak replied:
Yeah, wasn't sure about the name, but I clearly wanted to preserve some information that we are not making some span context a parent of a span that's about to be created because of the `WithNewRoot()` option. So I wanted some information that tells that this remote/local span context could be a parent, but isn't, because the code creating the span decided so. Not sure if `link-type` conveys that. `possible-remote-parent` or `possible-local-parent`? There are quite a mouthful already.

rghetia replied:
For now just open an issue.

krnowak replied:
Good idea, will do.

krnowak replied:
See #461.

krnowak commented:
Rebased.
